activerecord/lib/active_record/query_cache.rb

```diff
@@ -44,8 +44,9 @@ def self.install_executor_hooks(executor = ActiveSupport::Executor)
       executor.register_hook(self)
 
       executor.to_complete do
-        # FIXME: This should be skipped when env['rack.test']
-        ActiveRecord::Base.clear_active_connections!
+        unless ActiveRecord::Base.connection.transaction_open?
+          ActiveRecord::Base.clear_active_connections!
+        end
```

Comment:
I think it should be more like this, as ActiveRecord::Base.clear_active_connections! will clear current thread connection on all connection pool
```ruby
ActiveRecord::Base.connection_handler.connection_pool_list.each do |pool|
   pool.release_connection unless pool.active_connection? && pool.connection.transaction_open?
end
```